docs/4_same_classes.md

```diff
@@ -0,0 +1,41 @@
+
+Using course title alone would have not been enough, however. For instance, courses sometimes change their titles over the years, which would lead to our method failing to detect these. In addition, different courses are often offered under the same title even in the same semester. For instance, in Fall of 2020, Rick Antle taught "Foundations of Accounting and Valuation" as a Yale College course under **ACCT 270**, but Anya Nakhmurina taught a different course under the same name at the school of management, under the codes **MGT 502** and **HPM 502**.
+
+Using course description also introduces similar caveats as course titles do. In fact, professors are likely to make small modifications to their descriptions over time (for instance, changing the textbook used) that make exact matching suboptimal. Although we can use measures of fuzzy string similarity as a heuristic, we would ideally want to consider all three of code, title, and description when making our judgments.
+
+## Methodology
+
+Our input data source is a table of classes as well as one of listings – the `courses` and `listings` tables. The `courses` table stores a unique entry for each offering of a course, deduplicated by cross-listings (using cross-listing data provided by Yale), along with several attributes including title and description. The `listings` table stores the cross-listings of each course, which amounts to the code a course was listed in, as well as a pointer back to the deduplicated course in `courses`.
+
+The stated objective is to find all equivalent offerings for a given course. Since this amounts to evaluating an equivalence relationship on our set of courses, we can reformulate this task as computing a partition of our set of courses into groups of equivalent courses. Each group contains different offerings of the same course, and our set of groups is pairwise disjoint. Therefore, we just assign a unique group label to each course.
+
+Note that our previous strategy of identifying equivalent courses by shared code does not amount to an equivalence relationship – it's possible for courses *A*, B as well as *B*, *C* to have common codes but *A* and *C* to have no common ones, which violates the transitivity condition. This pattern sometimes occurs when course codes drift over time.
+
+To find these groups of equivalent courses, or equivalence classes, we model the set of courses as nodes in a graph, where each course is represented by a node. We then connect two nodes if and only if they are the same course. As a result, finding groups of equivalent courses amounts to finding [connected components](https://en.wikipedia.org/wiki/Component_(graph_theory)) in this graph, and it is easy to see that connected components are themselves a partition of the set of courses as desired. In fact, we previously used connected components when resolving cross-listings (sometimes Yale's cross-listings are incomplete).
```

Comment:
In this formulation, would it make sense to impose that an edge requires the two nodes to differ by a maximum amount of time, such as a couple of years. 

Consider course A, introduced in 2010 which mutates over time from x to y. If course B, also introduced in 2010, mutates from z to x, would the current method link A and B together? I believe it would be better to keep these two course chains separate.

Reply:
An interesting error case – however, wouldn't this requirement eliminate cases where a course is being offered again the next semester? To provide some additional context, we did take a look at whether or not course code reuse (i.e. a course code being recycled for a completely different course) was only happening after a set number of years, and it appears that reuse occurs as early as the next semester: https://github.com/coursetable/ferry/issues/54#issuecomment-727770209.

Also, I'm not sure if the example you provide is possible in practice – if course A is offered as x in 2010, then x cannot to refer to any future offering (each course offering per semester is stored as a separate instance in the `courses` table and therefore the graph). Therefore, it's not possible for course B to mutate to x post-2010.

Reply:
I think I wrote this before realizing we don't account for false negatives based on the title and description similarity. If we did account for false negatives, something like this could help avoid false false negatives. As long as the results look good right now, this would probably just add unnecessary complexity.

Reply:
Sounds good! Thanks for clarifying.